### PR TITLE
Add Shutdown RPC

### PR DESCRIPTION
**Pull request by vadmeste**
Implementation of minio control shutdown cmd with --restart flag

### CODE REVIEW

File: rpc-control.go [hunk 1]
```diff
@@ -26,13 +27,19 @@ import (
 const (
 	controlRPCPath = reservedBucket + "/control"
 	healPath       = controlRPCPath + "/heal"
+	shutdownPath   = controlRPCPath + "/shutdown"
 )
 
 // Register control RPC handlers.
 func registerControlRPCRouter(mux *router.Router, objAPI ObjectLayer) {
+	fmt.Printf("registering rpc\n")
```
Comment (harshavardhana) [inline rpc-control.go]: Remove 'fmt'

File: rpc-control.go [hunk 2]
```diff
@@ -26,13 +27,19 @@ import (
...
 	healRPCServer := rpc.NewServer()
 	healRPCServer.RegisterName("Heal", &healHandler{objAPI})
 	mux.Path(healPath).Handler(healRPCServer)
+
+	shutdownRPCServer := rpc.NewServer()
+	shutdownRPCServer.RegisterName("Shutdown", &shutdownHandler{})
```
Comment (harshavardhana) [inline rpc-control.go]: Let's not create new handlers for each. Move this to 'controllerAPIHandlers' and make this behave similar to web-router.go

File: rpc-control.go [hunk 3]
```diff
@@ -83,3 +90,25 @@ type HealObjectReply struct{}
...
+
+// ShutdownArgs - argument for Shutdown RPC.
+type ShutdownArgs struct {
+	Reboot bool
+}
+
+// ShutdownReply - reply by Shutdown RPC.
+type ShutdownReply struct{}
+
+// Shutdown - Shutdown the server.
```
Comment (harshavardhana) [inline rpc-control.go]: Move all these functions to 'controller-handlers.go'

File: control-shutdown.go [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+	"github.com/minio/cli"
+)
+
+var shutdownCmd = cli.Command{
+	Name:   "shutdown",
+	Usage:  "Shutdown or reboot the server.",
+	Action: shutdownControl,
+	Flags: []cli.Flag{
+		cli.BoolFlag{
+			Name:  "reboot",
```
Comment (harshavardhana) [inline control-shutdown.go]: This should be 'restart'

File: control-shutdown.go [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+USAGE:
+  minio control {{.Name}} http://localhost:9000/
+
+EAMPLES:
+  1. Shutdown the server:
+    $ minio control shutdown http://localhost:9000/
+
+  2. Reboot the server:
+    $ minio control shutdown --reboot http://localhost:9000/
```
Comment (harshavardhana) [inline control-shutdown.go]: Are you naming this based on `shutdown` unix command?

File: control-main.go [hunk 6]
```diff
@@ -33,102 +25,10 @@ var controlCmd = cli.Command{
 	Action: mainControl,
 	Subcommands: []cli.Command{
 		healCmd,
+		shutdownCmd,
 	},
```
Comment (harshavardhana) [inline control-main.go]: Add heal custom profile here look at mc `events-main.go`

File: rpc-control.go [hunk 7]
```diff
@@ -26,13 +27,19 @@ import (
 const (
 	controlRPCPath = reservedBucket + "/control"
 	healPath       = controlRPCPath + "/heal"
+	shutdownPath   = controlRPCPath + "/shutdown"
```
Comment (harshavardhana) [inline rpc-control.go]: Check the refactor here - https://github.com/minio/minio/commit/1d9617044348d2ab556eab005f55e7f8331e3553, once your patch is merged will rebase distributed branch properly.

File: cmd/controller-router.go [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+func registerControlRPCRouter(mux *router.Router, ctrlHandlers *controllerAPIHandlers) {
+	ctrlRPCServer := rpc.NewServer()
+	ctrlRPCServer.RegisterName("Control", ctrlHandlers)
+
+	ctrlRouter := mux.NewRoute().PathPrefix(reservedBucket).Subrouter()
+	ctrlRouter.Path(controlPath).Handler(ctrlRPCServer)
+}
+
+// Handler for object healing.
+type controllerAPIHandlers struct {
```
Comment (krishnasrinivas) [inline cmd/controller-router.go:40]: the name can be `controlAPIHandler` as we use the term `control` at other places and not `controller`

Comment (harshavardhana) [inline cmd/controller-router.go:40]: in lieu with above comment @vadmeste  can you rename '/control' as '/controller'

Comment (vadmeste) [inline cmd/controller-router.go:40]: :+1:

**Pull request merged**